File: cpp/dt_parser.c
#include <stdio.h>
#include <string.h>

char* prs_up_rpt_dt(char* s);
char* prs_dt_u(char* s); 
char* prs_up_rpt_dt_obj(char* id, char* s); 
char* prs_evt_ext(char* s);
char* prs_at_ext(char* s);

char* prs_up_rpt_dt(char* s) {
    static char dt[1024];
    memset(dt,0,sizeof(dt));
    int i=0;
    strcat(dt, "{\"cr\":\"");
    strncat(dt, s+i, 2);
    i+=2;
    strcat(dt, "\",\"ctrl\":\"");
    strncat(dt, s+i, 2);
    static char ctrl[3];
    memset(ctrl, 0, sizeof(ctrl));
    strncat(ctrl, s+i, 2);
    i+=2;
    strcat(dt, "\",\"l\":\"");
    strncat(dt, s+i, 2);
    i+=2;
    strcat(dt, "\",\"c\":\"");
    strncat(dt, s+i, 4);
    i+=4;
    strcat(dt, "\",\"r\":\"");
    strncat(dt, s+i, 4);
    i+=4;
    strcat(dt, "\",\"seq\":\"");
    strncat(dt, s+i, 2);
    static char seq[3];
    memset(seq,0,sizeof(seq));
    strncat(seq, s+i, 2);
    i+=2;
    printf("[seq]%s\n", seq);
    if(s[i-1]=='0') {
        strcat(dt, "\",\"appId\":\"");
        strncat(dt, s+i, 46);
        i+=46;
    }
    strcat(dt, "\",\"dt\":\"");
    strcat(dt, s+i);
    strcat(dt, "\"}");
    // return dt json
    printf("[dt]%s\n", dt);
    //return dt;
    if (ctrl[1] == '2' && seq[1] =='1') {
        static char dt_obj[512];
        memset(dt_obj,0,sizeof(dt));
        strcat(dt_obj, "{\"ext\":{\"dt\":\"");
        strcat(dt_obj, s+i);
        strcat(dt_obj, "\"}}");
        return dt_obj;
    }
    return prs_dt_u(s+i);
}

char* prs_dt_u(char* s) {
    static char dt_u[1024];
    int i=0;
    memset(dt_u,0,sizeof(dt_u));
    strcat(dt_u, "{\"ctrl\":\"");
    strncat(dt_u, s+i, 2);
    i+=2;
    strcat(dt_u, "\",\"n\":\"");
    strncat(dt_u, s+i, 2);
    i+=2;
    strcat(dt_u, "\",\"id\":\"");
    strncat(dt_u, s+i, 4);
    static char id[5];
    memset(id,0,sizeof(id));
    strncat(id, s+i, 4);
    i+=4;
    strcat(dt_u, "\",\"dt\":\"");
    strcat(dt_u, s+i);
    strcat(dt_u, "\"}");
    // return data union
    printf("[dt_u]%s\n", dt_u);
    //return dt_u;
    return prs_up_rpt_dt_obj(id, s+i);
}
    
char* prs_up_rpt_dt_obj(char* id, char* s) {
    static char dt[1024];
    int i=0;
    memset(dt,0,sizeof(dt));
    strcat(dt, "{\"time\":\"");
    strncat(dt, s+i, 12);
    i+=12;
    strcat(dt, "\",\"sts\":\"");
    strncat(dt, s+i, 8);
    i+=8;
    strcat(dt, "\",\"iVStd\":\"");
    strncat(dt, s+i, 16);
    i+=16;
    strcat(dt, "\",\"iV\":\"");
    strncat(dt, s+i, 16);
    i+=16;
    strcat(dt, "\",\"fVStd\":\"");
    strncat(dt, s+i, 16);
    i+=16;
    strcat(dt, "\",\"fV\":\"");
    strncat(dt, s+i, 16);
    i+=16;
    strcat(dt, "\",\"flwRtStd\":\"");
    strncat(dt, s+i, 8);
    i+=8;
    strcat(dt, "\",\"flwRt\":\"");
    strncat(dt, s+i, 8);
    i+=8;
    strcat(dt, "\",\"t\":\"");
    strncat(dt, s+i, 8);
    i+=8;
    strcat(dt, "\",\"p\":\"");
    strncat(dt, s+i, 8);
    i+=8;
    strcat(dt, "\",\"engSum\":\"");
    strncat(dt, s+i, 16);
    i+=16;
    strcat(dt, "\",\"engDst\":\"");
    strncat(dt, s+i, 8);
    i+=8;
    strcat(dt, "\",\"acsV\":\"");
    strncat(dt, s+i, 4);
    i+=4;
    strcat(dt, "\",\"vlt\":\"");
    strncat(dt, s+i, 4);
    i+=4;
    strcat(dt, "\",\"pwr\":\"");
    strncat(dt, s+i, 2);
    i+=2;
    strcat(dt, "\",\"nb\":\"");
    strncat(dt, s+i, 8);
    i+=8;
    strcat(dt, "\",\"ecl\":\"");
    strncat(dt, s+i, 4);
    i+=4;
    strcat(dt, "\",\"cid\":\"");
    strncat(dt, s+i, 12);
    i+=12;
    strcat(dt, "\",\"rnf\":\"");
    strncat(dt, s+i, 4);
    i+=4;
    printf("[obj_id]%s\n", id);
    strcat(dt, "\",\"ext\":");
    if (id[3] == '3') {
        strcat(dt, prs_evt_ext(s+i));
    } else if (id[3] == '2') {
        strcat(dt, prs_at_ext(s+i));
    }
    strcat(dt, "}");
    return dt;
} 
char* prs_evt_ext(char*s) {
    static char dt[1024];
    int i=0;
    memset(dt,0,sizeof(dt));
    strcat(dt, "{\"evt\":\"");
    strncat(dt, s+i, 2);
    i+=2;
    strcat(dt, "\",\"fvStdS1\":\"");
    strncat(dt, s+i, 16);
    i+=16;
    strcat(dt, "\",\"dt\":[");
    for(int j=0; j<3; j++) {
        char tmp[128];
        memset(tmp, 0, sizeof(tmp));
        if(j>0){
            strcat(tmp, ",");
        }
        strcat(tmp, "{\"date\":\"");
        strncat(tmp, s+i, 6);
        i+=6;
        strcat(tmp, "\",\"fVStd\":\"");
        strncat(tmp, s+i, 16);
        i+=16;
        strcat(tmp, "\",\"fV\":\"");
        strncat(tmp, s+i, 16);
        i+=16;
        strcat(tmp, "\"}");
        strcat(dt, tmp);
    }
    strcat(dt, "]}");
    return dt;

}
char* prs_at_ext(char *s) {
    static char dt[1024];
    int i=0;
    memset(dt, 0, sizeof(dt));
    strcat(dt, "{\"date\":\"");
    strncat(dt, s+i, 6);
    i+=6;
    strcat(dt, "\",\"dt\":\"");
    strcat(dt, s+i);
    strcat(dt, "\"}");
    return dt;
 }
int main(int argc, char* argv[]) {
    if(NULL == argv[1]) {
        printf("pls input dt\n");
        return -1;
    }
    printf("prs_dt, %s\n", argv[1]);
    printf("%s\n", prs_up_rpt_dt(argv[1]));
    return 0;
}

File: py/dt_parser.py
#!/usr/bin/python3
import argparse
import json

def prs_up_rpt_cmd(s):
    '''
    parse hex data in auto report command which is part of a frame starting with "0x88"
    :param s: a hex string
    :return: a json string
    '''
    dt = {}
    i=0
    dt['cr']=s[i:i+2]
    i+=2
    dt['ctrl']=s[i:i+2]
    i+=2
    dt['l']=s[i:i+2]
    i+=2
    dt['c']=s[i:i+4]
    i+=4
    dt['r']=s[i:i+4]
    i+=4
    dt['seq']=s[i:i+2]
    i+=2
    seq = int(dt['seq'],16)
    #print('seq', seq)
    if seq >= 0 or seq == int('80', 16):
        #print('hit')
        dt['appId']=s[i:i+46]
        i+=46
    dt['dt']=s[i:]
    return json.dumps(dt);
def prs_dt_u(is_nake_dt, s):
    '''
    parse data union
    :param is_nake_dt: True, not a data union; False, normal data union
    :param s: a hex string
    :return: a json string
    '''
    dt={}
    i=0
    if is_nake_dt:
        #print("nake_dt")
        dt['dt']=s
        return json.dumps(dt);
    dt['ctrl']=s[i:i+2]
    i+=2
    dt['n']=s[i:i+2]
    i+=2
    dt['id']=s[i:i+4]
    i+=4
    dt['dt']=s[i:]
    return json.dumps(dt);
def prs_up_rpt_dt_obj(is_nake_dt,id,s):
    '''
    parse auto report data obj
    :param is_nake_dt: True, not a data union; False, normal data union
    :param id: data object id
    :param s: a hex string
    :return: a json string
    '''
    dt={}
    i=0
    if is_nake_dt and id=='8002':
        dt['ext']=prs_at_ext(is_nake_dt, s[i:])
        return json.dumps(dt);
    dt['time']=s[i:i+12]
    i+=12
    dt['sts']=s[i:i+8]
    i+=8
    dt['iVStd']=s[i:i+16]
    i+=16
    dt['iV']=s[i:i+16]
    i+=16
    dt['fVStd']=s[i:i+16]
    i+=16
    dt['fV']=s[i:i+16]
    i+=16
    dt['flwRtStd']=s[i:i+8]
    i+=8
    dt['flwRt']=s[i:i+8]
    i+=8
    dt['t']=s[i:i+8]
    i+=8
    dt['p']=s[i:i+8]
    i+=8
    dt['engSum']=s[i:i+16]
    i+=16
    dt['engDst']=s[i:i+8]
    i+=8
    dt['acsV']=s[i:i+4]
    i+=4
    dt['vlt']=s[i:i+4]
    i+=4
    dt['pwr']=s[i:i+2]
    i+=2
    dt['nb']=s[i:i+8]
    i+=8
    dt['ecl']=s[i:i+4]
    i+=4
    dt['cid']=s[i:i+12]
    i+=12
    dt['rnf']=s[i:i+4]
    i+=4
    #dt['dt']=s[i:]
    if id == '8003':
        dt['ext']=prs_evt_ext(s[i:])
    elif id == '8002':
        dt['ext']=prs_at_ext(is_nake_dt, s[i:])
    return json.dumps(dt);
def prs_evt_ext(s):
    '''
    parse extend data in up evt report data
    :param s: a hex string
    :return: a json string
    '''
    dt={}
    i=0
    dt['evt']=s[i:i+2]
    i+=2
    dt['fvStdS1']=s[i:i+16]
    i+=16
    dt['dt']=[]
    for j in range(3):
        dt1={}
        dt1['date']=s[i:i+6]
        i+=6
        dt1['fVStd']=s[i:i+16]
        i+=16
        dt1['fV']=s[i:i+16]
        i+=16
        dt['dt'].append(dt1)
    return dt
def prs_at_ext(is_nake_dt, s):
    '''
    parse extend data in up auto report data
    :param s: a hex string
    :return: a json string
    '''
    dt={}
    i=0
    if is_nake_dt:
        dt['dt']=s[i:]
        return dt
    dt['date']=s[i:i+6]
    i+=6
    dt['dt']=s[i:]
    return dt
def prs_frm(s):
    '''
    parse hex data in frame start with "0x68"
    :param s:
    :return:
    '''
    return '';




if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("-c", "--cmd", type=str,help="输入明文指令hex字符串")
    parser.add_argument("-f", "--frame", type=str, help="输入明文帧hex字符串")

    args = parser.parse_args()
    cmd = args.cmd
    frame = args.frame
    if cmd:
        #print("cmd:", cmd)
        dt=prs_up_rpt_cmd(cmd)
        cmd_obj=json.loads(dt)
        #print("cmd_obj", cmd_obj)
        is_nake_dt = False
        if cmd_obj["ctrl"] == "C2" and cmd_obj["seq"] == "81":
            is_nake_dt = True
        dt=prs_dt_u(is_nake_dt, cmd_obj["dt"])
        obj=json.loads(dt)
        if is_nake_dt:
            obj["id"]='8002'
        dt=prs_up_rpt_dt_obj(is_nake_dt,obj["id"], obj["dt"])
    elif frame:
        #print("frame:", frame)
        json=prs_frm(frame)
    print(dt)
